nocturne/envs/vec_env_ma.py

```diff
@@ -0,0 +1,242 @@
+        self.frac_goal_achieved = []  # Log fraction of agents that achieved their goal
+        self.agents_in_scene = []
+
+    def _reset_seeds(self) -> None:
+        """Reset all environments' seeds."""
+        self._seeds = None
+
+    def reset(self, seed=None):
+        """Reset environment and return initial observations."""
+        # Reset Nocturne env
+        obs_dict = self.env.reset()
+
+        # Reset storage
+        self.agent_ids = []
+        self.rewards = []
+        self.dead_agent_ids = []
+        self.ep_collisions = 0
+        self.ep_goal_achived = 0
+
+        obs_all = np.full(fill_value=np.nan, shape=(self.num_envs, self.env.observation_space.shape[0]))
+        for idx, agent_id in enumerate(obs_dict.keys()):
+            self.agent_ids.append(agent_id)
+            obs_all[idx, :] = obs_dict[agent_id]
+
+        # Save obs in buffer
+        self._save_obs(obs_all)
+
+        logging.debug(f"RESET - agent ids: {self.agent_ids}")
+
+        # Make dict for storing the last info set for each agent
+        self.last_info_dicts = {agent_id: {} for agent_id in self.agent_ids}
+
+        return self._obs_from_buf()
+
+    def step(self, actions) -> VecEnvStepReturn:
+        """Convert action vector to dict and call env.step()."""
+
+        agent_actions = {
+            agent_id: actions[idx] for idx, agent_id in enumerate(self.agent_ids) if agent_id not in self.dead_agent_ids
+        }
+
+        # Take a step to obtain dicts
+        next_obses_dict, rew_dict, done_dict, info_dict = self.env.step(agent_actions)
+
+        # Update dead agents based on most recent done_dict
+        for agent_id, is_done in done_dict.items():
+            if is_done and agent_id not in self.dead_agent_ids:
+                self.dead_agent_ids.append(agent_id)
+                # Store agents' last info dict
+                self.last_info_dicts[agent_id] = info_dict[agent_id].copy()
+
+        # Convert dicts to arrays
+        obs_all = np.full(
+            fill_value=np.nan,
+            shape=(self.num_envs, self.env.observation_space.shape[0]),
+        )
+        rew_all = np.full(fill_value=np.nan, shape=(self.num_envs))
+        done_all = np.full(fill_value=np.nan, shape=(self.num_envs))
+        info_all = []
+
+        for idx, key in enumerate(self.agent_ids):
+            # Store data if available; otherwise leave as NaN
+            if key in next_obses_dict:
+                obs_all[idx, :] = next_obses_dict[key]
+                rew_all[idx] = rew_dict[key]
+                done_all[idx] = done_dict[key] * 1  # Will be 0 or 1 if valid, NaN otherwise
+
+        # OVERRIDE old buffer vals with with new ones (for all envs)
+        for env_idx in range(self.num_envs):
+            info_all.append(info_dict[key])
+            self.buf_rews[env_idx] = rew_all[env_idx]
+            self.buf_dones[env_idx] = done_all[env_idx]
+            self.buf_infos[env_idx] = info_all[env_idx]
```

Comment:
I think I'm slightly puzzled why there's two steps here where you store these values in the x_all objects and then copy them over to the buffer.

Reply:
good point, it's indeed redundant, I will compress L109 - L129